src/tactic/ring_exp.lean

```diff
@@ -0,0 +1,1430 @@
+
+  The `zero` constructor is the base case for `ex sum`, e.g. `1 + 2` is represented
+  by (something along the lines of) `sum 1 (sum 2 zero)`.
+
+  The `coeff` constructor is the base case for `ex prod`, and is used for numerals.
+  The code maintains the invariant that the coefficient is never `0`.
+
+  The `var` constructor is the base case for `ex exp`, and is used for atoms.
+
+  The `sum_b` constructor allows for addition in the base of an exponentiation;
+  it serves a similar purpose as the parentheses in `(a + b)^c`.
+  The code maintains the invariant that the argument to `sum_b` is not `zero`
+  or `sum _ zero`.
+
+  All of the constructors contain an `ex_info` field,
+  used to carry around (arguments to) proof terms.
+
+  While the `ex_type` parameter enforces some simplification invariants,
+  the following ones must be manually maintained at the risk of insufficient power:
+   - the argument to `coeff` must be nonzero (to ensure `0 = 0 * 1`)
+   - the argument to `sum_b` must be of the form `sum a (sum b bs)` (to ensure `(a + 0)^n = a^n`)
+   - normalisation proofs of subexpressions must be `refl ps.pretty`
+   - if we replace `sum` with `cons` and `zero` with `nil`, the resulting list is sorted
+     according to the `lt` relation defined further down; similarly for `prod` and `coeff`
+     (to ensure `a + b = b + a`).
+
+  The first two invariants could be encoded in a subtype of `ex`,
+  but aren't (yet) to spare some implementation burden.
+  The other invariants cannot be encoded because we need the `tactic` monad to check them.
+  (For example, the correct equality check of `expr` is `is_def_eq : expr → expr → tactic unit`.)
+-/
+meta inductive ex : ex_type → Type
+| zero  {} (info : ex_info) : ex sum
+| sum   {} (info : ex_info) : ex prod → ex sum → ex sum
+| coeff {} (info : ex_info) : coeff → ex prod
+| prod  {} (info : ex_info) : ex exp → ex prod → ex prod
+| var   {} (info : ex_info) : atom → ex base
+| sum_b {} (info : ex_info) : ex sum → ex base
+| exp   {} (info : ex_info) : ex base → ex prod → ex exp
+
+/--
+  Return the proof information associated to the `ex`.
+-/
+meta def ex.info : Π {et : ex_type} (ps : ex et), ex_info
+| sum  (ex.zero  i)     := i
+| sum  (ex.sum   i _ _) := i
+| prod (ex.coeff i _)   := i
+| prod (ex.prod  i _ _) := i
+| base (ex.var   i _)   := i
+| base (ex.sum_b i _)   := i
+| exp  (ex.exp   i _ _) := i
+
+/--
+  Return the original, non-normalized version of this `ex`.
+
+  Note that arguments to another `ex` are always "pre-normalized":
+  their `orig` and `pretty` are equal, and their `proof` is reflexivity.
+-/
+meta def ex.orig {et : ex_type} (ps : ex et) : expr := ps.info.orig
+/--
+  Return the normalized version of this `ex`.
+-/
+meta def ex.pretty {et : ex_type} (ps : ex et) : expr := ps.info.pretty
+/--
+  Return the normalisation proof of the given expression.
+  If the proof is `refl`, we give `none` instead,
```

Comment:
This is only done when the expression is syntactically equal to its normal form, right? You're not using `refl` to do any computation?